node/network/bridge/src/validator_discovery.rs

```diff
@@ -105,111 +59,64 @@ pub(super) struct Service<N, AD> {
 	/// and returns them as a workaround for the Future: Send requirement imposed by async fn impl.
 	pub async fn on_request(
 		&mut self,
 		validator_ids: Vec<AuthorityDiscoveryId>,
 		peer_set: PeerSet,
-		keep_alive: oneshot::Receiver<()>,
 		mut network_service: N,
 		mut authority_discovery_service: AD,
 	) -> (N, AD) {
-		const MAX_ADDR_PER_PEER: usize = 3;
-
-		let state = &mut self.state[peer_set];
-		// Increment the counter of how many times the validators were requested.
-		validator_ids.iter().for_each(|id| *state.requested_validators.entry(id.clone()).or_default() += 1);
-
 		// collect multiaddress of validators
-		let mut multiaddr_to_add = HashSet::new();
-		for authority in validator_ids.iter() {
+		let mut newly_requested = HashSet::new();
+		for authority in validator_ids.into_iter() {
 			let result = authority_discovery_service.get_addresses_by_authority_id(authority.clone()).await;
 			if let Some(addresses) = result {
-				// We might have several `PeerId`s per `AuthorityId`
-				multiaddr_to_add.extend(addresses.into_iter().take(MAX_ADDR_PER_PEER));
+				newly_requested.extend(addresses);
 			} else {
 				tracing::debug!(target: LOG_TARGET, "Authority Discovery couldn't resolve {:?}", authority);
 			}
 		}
 
+		let state = &mut self.state[peer_set];
 		// clean up revoked requests
-		let mut revoked_indices = Vec::new();
-		let mut revoked_validators = Vec::new();
-		for (i, maybe_revoked) in state.non_revoked_discovery_requests.iter_mut().enumerate() {
-			if maybe_revoked.is_revoked() {
-				for id in maybe_revoked.requested() {
-					if let Some(id) = on_revoke(&mut state.requested_validators, id.clone()) {
-						revoked_validators.push(id);
-					}
-				}
-				revoked_indices.push(i);
-			}
-		}
-
-		// clean up revoked requests states
-		//
-		// note that the `.rev()` here is important to guarantee `swap_remove`
-		// doesn't invalidate unprocessed `revoked_indices`
-		for to_revoke in revoked_indices.into_iter().rev() {
-			drop(state.non_revoked_discovery_requests.swap_remove(to_revoke));
-		}
-
-		// multiaddresses to remove
-		let mut multiaddr_to_remove = HashSet::new();
-		for id in revoked_validators.into_iter() {
-			let result = authority_discovery_service.get_addresses_by_authority_id(id.clone()).await;
-			if let Some(addresses) = result {
-				multiaddr_to_remove.extend(addresses.into_iter());
-			} else {
-				tracing::debug!(
-					target: LOG_TARGET,
-					"Authority Discovery couldn't resolve {:?} on cleanup, a leak is possible",
-					id,
-				);
-			}
-		}
+		let multiaddr_to_remove = state.previously_requested
+			.difference(&newly_requested)
+			.cloned()
+			.collect();
+		state.previously_requested = newly_requested.clone();
 
 		// ask the network to connect to these nodes and not disconnect
 		// from them until removed from the set
 		if let Err(e) = network_service.add_to_peers_set(
 			peer_set.into_protocol_name(),
-			multiaddr_to_add.clone(),
+			newly_requested,
```

Comment:
We will have duplicates here, which should be fine - just noting.

Reply:
this is a `HashSet`, do you duplicates from the previous request? that was the case before as well

Reply:
That's what I meant :-) Just mentioned it, because Pierre said somewhere before that peerset tracking (for debugging) is kind of hard in substrate because of duplicates. Either way, just forget about it.

Reply:
Fixed in d6d4870.